File: Android-App/app/src/main/java/com/example/myapplication/data/remote/Dtos.kt
package com.example.myapplication.data.remote

// Mirrors the JSON contracts exposed by the .NET API (System.Text.Json camelCase).

// ── Auth ─────────────────────────────────────────────
data class RegisterRequestDto(
    val fullName: String,
    val email: String,
    val password: String,
    val phone: String? = null,
    val plate: String? = null,
    val vehicleTypeId: Short? = null,
    val brand: String? = null,
    val model: String? = null,
    val color: String? = null,
)

data class LoginRequestDto(val email: String, val password: String)

data class AuthResponseDto(
    val userId: Int,
    val fullName: String,
    val email: String,
    val role: String,
    val municipalityId: Int?,
)

// ── Zones ────────────────────────────────────────────
data class ZoneDto(
    val id: Int,
    val municipalityId: Int,
    val municipalityName: String,
    val name: String,
    val description: String?,
    val latitude: Double,
    val longitude: Double,
    val totalSpots: Int,
    val freeSpots: Int,
    val hourlyRate: Double,
    val openHour: Int = 6,
    val closeHour: Int = 22,
    val isActive: Boolean,
)

data class CreateZoneRequestDto(
    val municipalityId: Int,
    val name: String,
    val description: String?,
    val latitude: Double,
    val longitude: Double,
    val totalSpots: Int,
    val hourlyRate: Double,
    val openHour: Int = 6,
    val closeHour: Int = 22,
)

data class UpdateZoneRequestDto(
    val name: String,
    val description: String?,
    val totalSpots: Int,
    val hourlyRate: Double,
    val openHour: Int,
    val closeHour: Int,
    val isActive: Boolean,
)

// ── Vehicles ─────────────────────────────────────────
data class VehicleDto(
    val id: Int,
    val userId: Int,
    val vehicleTypeId: Int,
    val vehicleType: String,
    val plate: String,
    val nickname: String?,
    val brand: String?,
    val model: String?,
    val color: String?,
    val isActive: Boolean,
)

data class CreateVehicleRequestDto(
    val userId: Int,
    val vehicleTypeId: Short,
    val plate: String,
    val nickname: String? = null,
    val brand: String? = null,
    val model: String? = null,
    val color: String? = null,
)

// ── Payment methods ──────────────────────────────────
data class PaymentMethodDto(
    val id: Int,
    val cardBrand: String?,
    val lastFour: String,
    val expiryMonth: Short,
    val expiryYear: Short,
    val isDefault: Boolean,
)

data class CreatePaymentMethodRequestDto(
    val userId: Int,
    val cardBrand: String?,
    val lastFour: String,
    val expiryMonth: Short,
    val expiryYear: Short,
    val token: String,
    val isDefault: Boolean,
)

// ── Sessions ─────────────────────────────────────────
data class CreateSessionRequestDto(
    val userId: Int,
    val vehicleId: Int,
    val zoneId: Int,
    val spaceNumber: Int,
    val scheduledStart: String,
    val scheduledEnd: String,
)

data class CreateSessionResponseDto(val id: Int, val totalCost: Double)

data class ExtendSessionRequestDto(val addedMinutes: Int)

data class ExtendSessionResponseDto(
    val id: Int,
    val scheduledEnd: String,
    val totalCost: Double,
    val additionalCost: Double,
    val addedMinutes: Int,
)

data class FinalizeSessionResponseDto(val id: Int, val status: String, val totalCost: Double)

data class SessionDto(
    val id: Int,
    val zoneId: Int,
    val zoneName: String,
    val vehicleId: Int,
    val plate: String,
    val spaceNumber: Int,
    val scheduledStart: String,
    val scheduledEnd: String,
    val actualEnd: String?,
    val hourlyRate: Double,
    val totalCost: Double,
    val status: String,
    val zoneOpenHour: Int = 6,
    val zoneCloseHour: Int = 22,
)

// ── Payments ─────────────────────────────────────────
data class CreatePaymentRequestDto(
    val userId: Int,
    val paymentMethodId: Int?,
    val amount: Double,
    val referenceType: String,
    val referenceId: Int,
)

data class PaymentDto(
    val id: Int,
    val amount: Double,
    val currency: String,
    val status: String,
    val referenceType: String,
    val referenceId: Int,
    val paidAt: String?,
    val invoiceNumber: String?,
)

// ── Fines ────────────────────────────────────────────
data class FineDto(
    val id: Int,
    val vehicleId: Int,
    val plate: String,
    val zoneId: Int,
    val zoneName: String,
    val spaceNumber: Int? = null,
    val reason: String,
    val evidenceUrl: String?,
    val amount: Double,
    val status: String,
    val issuedAt: String,
    val paidAt: String?,
)

data class CreateFineRequestDto(
    val issuedBy: Int,
    val vehicleId: Int,
    val zoneId: Int,
    val reason: String,
    val evidenceUrl: String?,
    val amount: Double,
)

// ── Reports ──────────────────────────────────────────
data class ReportSummaryDto(
    val totalSessions: Int,
    val revenue: Double,
    val finesIssued: Int,
    val activeSpots: Int,
    val totalSpots: Int,
)

File: api/eparkapi/Models/Dtos/Dtos.cs
namespace eparkapi.Models.Dtos;

// ── Auth ─────────────────────────────────────────────
public record RegisterRequest(
    string FullName,
    string Email,
    string Password,
    string? NationalId,
    string? Phone,
    string? Plate,
    short? VehicleTypeId,
    string? Brand,
    string? Model,
    string? Color);

public record LoginRequest(string Email, string Password);

public record ResendVerificationRequest(string Email);

public record AuthResponse(int UserId, string FullName, string Email, string Role, int? MunicipalityId);

// ── Zones ────────────────────────────────────────────
public record ZoneResponse(
    int Id,
    int MunicipalityId,
    string MunicipalityName,
    string Name,
    string? Description,
    decimal Latitude,
    decimal Longitude,
    int TotalSpots,
    int FreeSpots,
    decimal HourlyRate,
    int OpenHour,
    int CloseHour,
    bool IsActive);

public record CreateZoneRequest(
    int MunicipalityId,
    string Name,
    string? Description,
    decimal Latitude,
    decimal Longitude,
    int TotalSpots,
    decimal HourlyRate,
    int OpenHour = 6,
    int CloseHour = 22);

public record UpdateZoneRequest(
    string Name,
    string? Description,
    int TotalSpots,
    decimal HourlyRate,
    int OpenHour,
    int CloseHour,
    bool IsActive);

// ── Vehicles ─────────────────────────────────────────
public record VehicleResponse(
    int Id,
    int UserId,
    int VehicleTypeId,
    string VehicleType,
    string Plate,
    string? Nickname,
    string? Brand,
    string? Model,
    string? Color,
    bool IsActive);

public record CreateVehicleRequest(
    int UserId,
    short VehicleTypeId,
    string Plate,
    string? Nickname,
    string? Brand,
    string? Model,
    string? Color);

// ── Payment methods ──────────────────────────────────
public record PaymentMethodResponse(
    int Id,
    string? CardBrand,
    string LastFour,
    short ExpiryMonth,
    short ExpiryYear,
    bool IsDefault);

public record CreatePaymentMethodRequest(
    int UserId,
    string? CardBrand,
    string LastFour,
    short ExpiryMonth,
    short ExpiryYear,
    string Token,
    bool IsDefault);

// ── Sessions ─────────────────────────────────────────
public record CreateSessionRequest(
    int UserId,
    int VehicleId,
    int ZoneId,
    int SpaceNumber,
    DateTime ScheduledStart);

public record ExtendSessionRequest(int AddedMinutes);

public record SessionResponse(
    int Id,
    int ZoneId,
    string ZoneName,
    int VehicleId,
    string Plate,
    int SpaceNumber,
    DateTime ScheduledStart,
    DateTime ScheduledEnd,
    DateTime? ActualEnd,
    decimal HourlyRate,
    decimal TotalCost,
    string Status,
    int ZoneOpenHour,
    int ZoneCloseHour);

public record FinalizeSessionResponse(int Id, string Status, decimal TotalCost);

// ── Payments ─────────────────────────────────────────
public record CreatePaymentRequest(
    int UserId,
    int? PaymentMethodId,
    decimal Amount,
    string ReferenceType,
    int ReferenceId);

public record PaymentResponse(
    int Id,
    decimal Amount,
    string Currency,
    string Status,
    string ReferenceType,
    int ReferenceId,
    DateTime? PaidAt,
    string? InvoiceNumber);

// ── Fines ────────────────────────────────────────────
public record FineResponse(
    int Id,
    int VehicleId,
    string Plate,
    int ZoneId,
    string ZoneName,
    int SpaceNumber,
    string Reason,
    string? EvidenceUrl,
    decimal Amount,
    string Status,
    DateTime IssuedAt,
    DateTime? PaidAt);

public record CreateFineRequest(
    int IssuedBy,
    int VehicleId,
    int ZoneId,
    int SpaceNumber,
    string Reason,
    string? EvidenceUrl,
    decimal Amount);

// ── Reports ──────────────────────────────────────────
public record ZoneRevenueResponse(string ZoneName, decimal Revenue, int Sessions);

public record ReportSummaryResponse(
    int TotalSessions,
    decimal Revenue,
    int FinesIssued,
    int ActiveSpots,
    int TotalSpots,
    IEnumerable<ZoneRevenueResponse> RevenueByZone);

// ── Admin notifications ──────────────────────────────
// Derived live feed (no table): overdue active sessions + recently paid fines.
public record AdminNotificationResponse(
    string Id,
    string Type,
    int ReferenceId,
    string Plate,
    string ZoneName,
    DateTime Timestamp,
    decimal? Amount,
    int? SpaceNumber);

// ── Admin action log ─────────────────────────────────
public record AdminActionLogResponse(
    long Id,
    int AdminId,
    string AdminName,
    string Action,
    string? Details,
    DateTime CreatedAt);
